Use WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<> more in WebCore

ddkilzer opened the pull request:
#### 8f33f11f2a7f6cf72197a8651b6a9cf17136154c
<pre>
Use WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr&lt;&gt; more in WebCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=250929">https://bugs.webkit.org/show_bug.cgi?id=250929</a>
&lt;rdar://problem/104436726&gt;

Reviewed by Chris Dumez.

Classes that inherit from both WTF::ThreadSafeRefCounted&lt;&gt; and
WTF::CanMakeWeakPtr&lt;&gt; should use
WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr&lt;&gt; instead.

Most of the changes below are:
- to replace WeakPtr&lt;&gt; with ThreadSafeWeakPtr&lt;&gt; in instance
  variables and data structures,
- to change raw pointers returned from WeakPtr&lt;&gt;.get() to
  RefPtr&lt;&gt; types, and
- to use a local RefPtr&lt;&gt; to an object instead of checking the
  pointer value in the ThreadSafeWeakPtr&lt;&gt; object directly.

Since a lock is taken each time the value stored in the
ThreadSafeWeakPtr&lt;&gt; is returned, there is an effort to minimize
the number of times that happens within a given method by
creating a local RefPtr&lt;&gt; variable on the stack.

* Source/WebCore/platform/graphics/DecomposedGlyphs.h:
* Source/WebCore/platform/graphics/ImageSource.cpp:
(WebCore::ImageSource::ensureDecoderAvailable):
* Source/WebCore/platform/graphics/ImageSource.h:
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.h:
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.mm:
(-[WebAVPlayerViewControllerDelegate fullscreenInterface]):
(-[WebAVPlayerViewControllerDelegate setFullscreenInterface:]):
(-[WebAVPlayerViewControllerDelegate playerViewControllerWillStartPictureInPicture:]):
(-[WebAVPlayerViewControllerDelegate playerViewControllerDidStartPictureInPicture:]):
(-[WebAVPlayerViewControllerDelegate playerViewController:failedToStartPictureInPictureWithError:]):
(-[WebAVPlayerViewControllerDelegate playerViewControllerWillStopPictureInPicture:]):
(-[WebAVPlayerViewControllerDelegate playerViewControllerDidStopPictureInPicture:]):
(-[WebAVPlayerViewControllerDelegate playerViewController:shouldExitFullScreenWithReason:]):
(-[WebAVPlayerViewControllerDelegate playerViewController:restoreUserInterfaceForPictureInPictureStopWithCompletionHandler:]):
(-[WebAVPlayerViewControllerDelegate pictureInPictureControllerWillStartPictureInPicture:]):
(-[WebAVPlayerViewControllerDelegate pictureInPictureControllerDidStartPictureInPicture:]):
(-[WebAVPlayerViewControllerDelegate pictureInPictureController:failedToStartPictureInPictureWithError:]):
(-[WebAVPlayerViewControllerDelegate pictureInPictureControllerWillStopPictureInPicture:]):
(-[WebAVPlayerViewControllerDelegate pictureInPictureControllerDidStopPictureInPicture:]):
(-[WebAVPlayerViewControllerDelegate pictureInPictureController:restoreUserInterfaceForPictureInPictureStopWithCompletionHandler:]):
(-[WebAVPlayerViewController initWithFullscreenInterface:]):
(-[WebAVPlayerViewController enterFullScreenAnimated:completionHandler:]):
- On watchOS, if _fullscreenInterface is nullptr, attempt to
  call completionHandler if set.  Previously this would crash.
(-[WebAVPlayerViewController MY_NO_RETURN]):
* Source/WebCore/platform/network/cocoa/RangeResponseGenerator.h:
* Source/WebCore/platform/network/cocoa/RangeResponseGenerator.mm:
(WebCore::RangeResponseGenerator::giveResponseToTaskIfBytesInRangeReceived):
* Source/WebCore/storage/StorageQuotaManager.h:

* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::clearNativeImageMap):
(WebKit::RemoteResourceCacheProxy::clearDecomposedGlyphsMap):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:

Canonical link: <a href="https://commits.webkit.org/260107@main">https://commits.webkit.org/260107@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0f9950160535ef5326cb1c1e740f6e46ac0187fa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107052 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16105 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39844 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116228 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115738 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110953 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17562 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7301 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99229 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112817 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13279 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96316 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40886 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95171 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27953 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82637 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9208 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29304 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9805 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6307 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15367 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48870 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6988 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11347 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -183,15 +183,15 @@ FloatRect ImageBuffer::clampedRect(const FloatRect& rect)
-        if (source->hasOneRef())
+        if (source->hasOneStrongRef())
             return source;
     }
```
I'm not thrilled about this because it could be racy—for example, if `source` has its refcount increased between the `hasOneStrongRef()` check and returning from the function—but I'm not sure it's problem in practice.

To be fair, since `source` was previously `ThreadSafeRefCounted<>`, this change basically has the same behavior as the previous code.

@shallawa Do you have any feedback here?  For example, if `source` has a refcount of 1, does that infer that it is not being used on other threads, too?

ddkilzer replied:
Moved `WebCore::ImageBuffer` to another bug/patch.

cdumez commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -136,6 +143,8 @@ class ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr {
 public:
     void ref() const { m_controlBlock.strongRef(); }
     void deref() const { m_controlBlock.template strongDeref<destructionThread>(); }
+    bool hasOneStrongRef() const { return m_controlBlock.strongRefCount() == 1;  }
```
I seem to remember @achristensen07 purposefully didn’t add this one because it is racy?

ddkilzer replied:
See my [comment](https://github.com/WebKit/WebKit/pull/8915/files#r1083243910) in `ImageBuffer.cpp` below.

The `.hasOneRef()` was racy before, so there's technically no behavior change, but this is less-than-ideal.

ddkilzer replied:
Woke up early this morning and a possible solution popped into my head.

Could we create a template function like this:

```
template <typename T, typename U>
U ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<T>::runWithRefCount(lambda_expr(size_t refCount) -> U);
```

This method would take the lock, then run _lambda_expr_ with the current ref count.

For the case in `ImageBuffer.cpp`, though, I'm not sure if this is really protecting anything.

There are a couple more instances where this is needed in debug assertions that are just `ASSERT(hasOneRef())`.  Maybe we need to figure out why `hasOneRef()` is a special state for this code.  Do you know @shallawa?

ddkilzer replied:
Moving `WebCore::ImageBuffer` changes to another bug/patch.

cdumez commented on Source/WebCore/platform/encryptedmedia/CDMProxy.cpp:
```diff
@@ -357,7 +357,7 @@ void CDMInstanceProxy::startedWaitingForKey()
     m_numDecryptorsWaitingForKey++;
 
-    callOnMainThread([player = m_player, wasWaitingForKey] {
+    callOnMainThread([player = m_player.get(), wasWaitingForKey] {
```
This is a behavior change, you are now keeping the player alive in the lambda. Why aren’t you capturing the weak pointer like we used to?

ddkilzer replied:
If we switch it back to a weak reference in the lambda, we just have to grab the strong reference on the main thread inside the lambda:

```
    callOnMainThread([weakPlayer = m_player, wasWaitingForKey] {
        auto player = weakPlayer.get();
        if (player && !wasWaitingForKey)
             player->waitingForKeyChanged();
     });
```

I'll make that change.

ddkilzer replied:
Fixed this in the second patch.

cdumez commented on Source/WebCore/platform/encryptedmedia/CDMProxy.cpp:
```diff
@@ -371,7 +371,7 @@ void CDMInstanceProxy::stoppedWaitingForKey()
     bool isNobodyWaitingForKey = !m_numDecryptorsWaitingForKey;
 
-    callOnMainThread([player = m_player, isNobodyWaitingForKey] {
+    callOnMainThread([player = m_player.get(), isNobodyWaitingForKey] {
```
Ditto.

ddkilzer replied:
Fixed this in the second patch.

cdumez commented on Source/WebCore/Modules/webxr/WebXRSystem.cpp:
```diff
@@ -558,10 +559,10 @@ void WebXRSystem::unregisterSimulatedXRDeviceForTesting(PlatformXR::Device& devi
-    ASSERT_UNUSED(removed, removed || m_inlineXRDevice == &device);
-    if (m_activeImmersiveDevice == &device)
+    ASSERT_UNUSED(removed, removed || m_inlineXRDevice.get().get() == &device);
+    if (m_activeImmersiveDevice.get().get() == &device)
```
Do we really need the second .get()? It is a bit ugly.

ddkilzer replied:
Ah, `RefPtr<>::operator==()` is overloaded to compare to a raw pointer, so we can remove the second `.get()`.

ddkilzer replied:
Will fix in the third patch (about to upload).

cdumez reviewed (commented):
Looks pretty good overall. Just a little concerned about the hasOneStrongRef() code. I will let Alex take a look.

cdumez reviewed (approved):
LGTM % nits.

ddkilzer commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -141,7 +141,12 @@ class ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr {
+    ThreadSafeWeakPtrControlBlock<T>& m_controlBlock { *new ThreadSafeWeakPtrControlBlock<T>(reinterpret_cast<T&>(*this)) };
+#else
     ThreadSafeWeakPtrControlBlock<T>& m_controlBlock { *new ThreadSafeWeakPtrControlBlock<T>(static_cast<T&>(*this)) };
+#endif
```
This MSVC workaround made WebKit compile, but I'm not sure it's correct.  Seeing some strange crashes on the layout tests from the [win](https://ews-build.webkit.org/#/builders/10/builds/174192) bot, though even the set of tests that are crashing is inconsistent.

Waiting for layout tests to run without the patch to compare.

cdumez replied:
Yes, I am not convinced this is correct. @achristensen07 can confirm but I believe he previously wrote a patch that updated these static_casts with reinterpret_casts and it got reverted because it introduced crashes.

ddkilzer replied:
Yep.  I found a better workaround.  Trying it now (once I fix my silly typos).

ddkilzer replied:
Arg.  Looking at the wrong build results.  This patch still needs a bit of work.

ddkilzer replied:
The latest patch adds a `DestructionThread` template parameter (with a default value) to both `WTF::ThreadSafeWeakPtrControlBlock<>` and `WTF::ThreadSafeWeakPtrControlBlock<>` so that all three classes have consistent template parameters.

This compiles on macOS and iOS—just waiting for Win/WinCairo ports to compile.

ddkilzer commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -85,17 +84,21 @@ class ThreadSafeWeakPtrControlBlock {
-            ensureOnMainRunLoop(WTFMove(deleteObject));
             break;
         }
+#endf
```
Typo:  `#endif`

Commit:
Use WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<> more in WebCore
https://bugs.webkit.org/show_bug.cgi?id=250929
<rdar://problem/104436726>

Reviewed by Chris Dumez.

Classes that inherit from both WTF::ThreadSafeRefCounted<> and
WTF::CanMakeWeakPtr<> should use
WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<> instead.

Most of the changes below are:
- to replace WeakPtr<> with ThreadSafeWeakPtr<> in instance
  variables and data structures,
- to change raw pointers returned from WeakPtr<>.get() to
  RefPtr<> types, and
- to use a local RefPtr<> to an object instead of checking the
  pointer value in the ThreadSafeWeakPtr<> object directly.

Since a lock is taken each time the value stored in the
ThreadSafeWeakPtr<> is returned, there is an effort to minimize
the number of times that happens within a given method by
creating a local RefPtr<> variable on the stack.

* Source/WebCore/platform/graphics/DecomposedGlyphs.h:
* Source/WebCore/platform/graphics/ImageSource.cpp:
(WebCore::ImageSource::ensureDecoderAvailable):
* Source/WebCore/platform/graphics/ImageSource.h:
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.h:
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.mm:
(-[WebAVPlayerViewControllerDelegate fullscreenInterface]):
(-[WebAVPlayerViewControllerDelegate setFullscreenInterface:]):
(-[WebAVPlayerViewControllerDelegate playerViewControllerWillStartPictureInPicture:]):
(-[WebAVPlayerViewControllerDelegate playerViewControllerDidStartPictureInPicture:]):
(-[WebAVPlayerViewControllerDelegate playerViewController:failedToStartPictureInPictureWithError:]):
(-[WebAVPlayerViewControllerDelegate playerViewControllerWillStopPictureInPicture:]):
(-[WebAVPlayerViewControllerDelegate playerViewControllerDidStopPictureInPicture:]):
(-[WebAVPlayerViewControllerDelegate playerViewController:shouldExitFullScreenWithReason:]):
(-[WebAVPlayerViewControllerDelegate playerViewController:restoreUserInterfaceForPictureInPictureStopWithCompletionHandler:]):
(-[WebAVPlayerViewControllerDelegate pictureInPictureControllerWillStartPictureInPicture:]):
(-[WebAVPlayerViewControllerDelegate pictureInPictureControllerDidStartPictureInPicture:]):
(-[WebAVPlayerViewControllerDelegate pictureInPictureController:failedToStartPictureInPictureWithError:]):
(-[WebAVPlayerViewControllerDelegate pictureInPictureControllerWillStopPictureInPicture:]):
(-[WebAVPlayerViewControllerDelegate pictureInPictureControllerDidStopPictureInPicture:]):
(-[WebAVPlayerViewControllerDelegate pictureInPictureController:restoreUserInterfaceForPictureInPictureStopWithCompletionHandler:]):
(-[WebAVPlayerViewController initWithFullscreenInterface:]):
(-[WebAVPlayerViewController enterFullScreenAnimated:completionHandler:]):
- On watchOS, if _fullscreenInterface is nullptr, attempt to
  call completionHandler if set.  Previously this would crash.
(-[WebAVPlayerViewController MY_NO_RETURN]):
* Source/WebCore/platform/network/cocoa/RangeResponseGenerator.h:
* Source/WebCore/platform/network/cocoa/RangeResponseGenerator.mm:
(WebCore::RangeResponseGenerator::giveResponseToTaskIfBytesInRangeReceived):
* Source/WebCore/storage/StorageQuotaManager.h:

* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::clearNativeImageMap):
(WebKit::RemoteResourceCacheProxy::clearDecomposedGlyphsMap):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:

Canonical link: https://commits.webkit.org/260107@main
